### PR TITLE
Initial implementations of a Special Attack

### LINKED ISSUE #115: Feat: Special, single-use move to wipe out the visible enemies
I think I want to resurrect this idea:

> Or, player could use a special key combo. which can only be used once in a single play (like Z-moves, if you know pokemon...), and it will cause all the enemies/trashes on the screen to be killed, yielding very large number of points at once. This could be used when the player sees a large number of trashes on-screen...

Sounds like something possible!

_Originally posted by @DiddiLeija in https://github.com/DiddiLeija/diddi-and-the-bugs/issues/13#issuecomment-1483701365_

### PR DESCRIPTION
**Pull request by HarshNarayanJha**
This is the initial setup for our Special Attack!
This special attack will destroy all trash/enemies currently on screen. It will also cut the Monster’s health to half, if currently shown. To do this I needed to make the monster die after two hits, using a new variable on it called `hit_count`.

Right now the attack is triggered using the Z key (an attribute to the inspiration of this Easter egg, Z-move in Pokémon).

cc @DiddiLeija, as you already know, I can’t test this code. Please try this out and let me know!

### CODE REVIEW

**Comment by DiddiLeija** [comment]
Great! I'll give it a look today :+1:

File: main.py [hunk 1]
```diff
@@ -177,6 +187,9 @@ def bullet_collision(self, bullet):
             )
             and bullet.alive
         )
+    
+    def hit_special_move(self):
+        self.hit_count += 1
```
Comment (DiddiLeija) [inline main.py:192]: Here's a whitespace, and seems like Flake8 is unhappy with that.

```suggestion

    def hit_special_move(self):
        self.hit_count += 1

```

File: main.py [hunk 2]
```diff
@@ -364,6 +378,18 @@ def update_game(self):
...
+            self.used_special_move = True
+            # some cool animations
+            for enem in self.enemies:
+                if enem.is_alive and enem.show:
+                    enem.hit_special_move()
+
+            if self.monster.show and self.monster.alive:
+                self.monster.hit_special_move()
+
+
```
Comment (DiddiLeija) [inline main.py:392]: Here, Flake8 proposes 2 faults: `multiple statements on one line (colon)`, which can be fixed easily in this case, and `too many blank lines (2)`. I propose this change:

```suggestion
        if pyxel.btnr(pyxel.KEY_Z):
            if self.used_special_move:
                return
            self.used_special_move = True
            # some cool animations
            for enem in self.enemies:
                if enem.is_alive and enem.show:
                    enem.hit_special_move()

            if self.monster.show and self.monster.alive:
                self.monster.hit_special_move()

```

**Review by DiddiLeija — COMMENTED** [review]
Great start! Thank you @HarshNarayanJha. I haven't tested the behavior yet, but in terms of code sanity... here're a few suggestions ;)

File: main.py [hunk 3]
```diff
@@ -364,6 +378,18 @@ def update_game(self):
             self.bullet_last_held_long = False
             self.continous_bullets_message = False
 
+        if pyxel.btnr(pyxel.KEY_Z):
+            if self.used_special_move: return
+            self.used_special_move = True
+            # some cool animations
+            for enem in self.enemies:
+                if enem.is_alive and enem.show:
```
Comment (DiddiLeija) [inline main.py:386]: It seems to me that `is_alive` was actually `alive`?

```suggestion
                if enem.alive and enem.show:
```

**Review by DiddiLeija — CHANGES_REQUESTED** [review]
Ok, so... I tested the behavior. Pretty nice! Though it seems like trash is unaffected by the attack (was that intentional? maybe).

I just found a small typo, that was uncatched by linters :upside_down_face:

**Comment by HarshNarayanJha** [comment]
> Though it seems like trash is unaffected by the attack (was that intentional? maybe).

No it wasn’t intentional.. as I forgot to iterate over trash list and kill them all 😅.. but I think it should be like that… only destroying enemies.

Now I only need to calculate the scores and add them… but what about animation and sound effect(s), we need something for it right?

**Comment by DiddiLeija** [comment]
> No it wasn’t intentional.. as I forgot to iterate over trash list and kill them all 😅.. but I think it should be like that… only destroying enemies.

Yeah, since trash is a bonus item (not needed to win) :wink:

> but what about animation and sound effect(s), we need something for it right?

Yup. Maybe I can give it a look.

Or, you can design an animation/sound track using https://kitao.github.io/pyxel/wasm/examples/image_editor.html. Then you can download the content, send it to me, and I'll paste it in our resource file once we merge this PR?

**Comment by HarshNarayanJha** [comment]
> Or, you can design an animation/sound track using

I really don’t know pixel art + I am currently learning animation (as you already know) and I really don’t know anything about music making. So sorry 😢

**Comment by DiddiLeija** [comment]
> > Or, you can design an animation/sound track using
> 
> I really don’t know pixel art + I am currently learning animation (as you already know) and I really don’t know anything about music making. So sorry 😢

Ok, don't worry. Gonna try to do it myself, at least in the artistic part :)

**Comment by HarshNarayanJha** [comment]
My implementations are done now. Opened for merging

**Review by DiddiLeija — APPROVED** [review]
It seems to be working, great job!
I'll try to get into graphics (and possibly docs) in the next days [^1]. ;)

[^1]: Oh, btw, maybe this PR should not close #115. I'll edit the first post in this thread :)

**Pull request merged**